### PR TITLE
CUSTOM KEYBINDS for all slots

### PR DESCRIPTION
**Pull request by fernandos13**
## About The Pull Request
fixes the #5807
does at it says give keybinds for EACH slots that you got. jjust like the old preferred keybind 
that would let you choose only ONE slot to pull from. now you can have multiple keys to multiple slots
image

## Why It's Good For The Game
QoL

## Changelog
:cl:
code: QoL adds keybinds to all slots. customize it as you like in keybind prefs
fix: delay for using hotkeys
/:cl:

### CODE REVIEW

**Comment by SkeletalElite** [comment]
Yes please, will be super nice for quick pulling ammo when I don't have ammo stored in my belt while still having E equip and unequip my gun.

**Comment by joookschad** [comment]
god YES

**Comment by fernandos13** [comment]
do not merge. in understand. but why the do not TM?

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 1]
```diff
@@ -31,6 +31,63 @@
...
+/mob/living/carbon/human/proc/do_kb_equip_slot_r_store()
+	do_kb_equip(SLOT_R_STORE)
+/mob/living/carbon/human/proc/do_kb_equip_slot_accessory()
+	do_kb_equip(SLOT_IN_ACCESSORY)
+/mob/living/carbon/human/proc/do_kb_equip_slot_beltholster()
+	do_kb_equip(SLOT_IN_HOLSTER)
+/mob/living/carbon/human/proc/do_kb_equip_slot_suitsholster()
+	do_kb_equip(SLOT_IN_S_HOLSTER)
+/mob/living/carbon/human/proc/do_kb_equip_slot_backholster()
+	do_kb_equip(SLOT_IN_B_HOLSTER)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:57]: use spaces

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 2]
```diff
@@ -31,6 +31,63 @@
...
+
+/mob/living/carbon/human/proc/do_kb_equip( KB )
+	SIGNAL_HANDLER_DOES_SLEEP
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:60]: This is not allowed

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 3]
```diff
@@ -31,6 +31,63 @@
...
+			return
+		if( KB )
+			if(equip_to_slot_if_possible(I, KB , FALSE, FALSE, FALSE))
+				return
+		if(!equip_to_appropriate_slot(I, FALSE))
+			return
+		if(hand)
+			update_inv_l_hand(FALSE)
+		else
+			update_inv_r_hand(FALSE)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:90]: use early returns

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 4]
```diff
@@ -31,6 +31,63 @@
...
+	. = COMSIG_KB_ACTIVATED //The return value must be a flag compatible with the signals triggering this.
+
+	if(incapacitated() || lying_angle || istype(loc, /obj/vehicle/multitile/root/cm_armored))
+		return
+
+	var/obj/item/I = get_active_held_item()
+	if(!I)
+		if(next_move > world.time)
+			return
+		if( KB )
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:70]: what is this wierd spacing

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 5]
```diff
@@ -31,6 +31,63 @@
...
+			if(draw_from_slot_if_possible( KB ))
+				next_move = world.time + 3
+				return
+		for(var/slot in SLOT_DRAW_ORDER)
+			if(draw_from_slot_if_possible(slot))
+				next_move = world.time + 3
+				return
+	else
+		if(s_active && s_active.can_be_inserted(I))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:79]: Use ?.

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 6]
```diff
@@ -31,6 +31,63 @@
...
+/mob/living/carbon/human/proc/do_kb_equip_slot_accessory()
+	do_kb_equip(SLOT_IN_ACCESSORY)
+/mob/living/carbon/human/proc/do_kb_equip_slot_beltholster()
+	do_kb_equip(SLOT_IN_HOLSTER)
+/mob/living/carbon/human/proc/do_kb_equip_slot_suitsholster()
+	do_kb_equip(SLOT_IN_S_HOLSTER)
+/mob/living/carbon/human/proc/do_kb_equip_slot_backholster()
+	do_kb_equip(SLOT_IN_B_HOLSTER)
+
+/mob/living/carbon/human/proc/do_kb_equip( KB )
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:59]: more wierd spacing and single letter vars

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 7]
```diff
@@ -31,6 +31,63 @@
...
+		if(next_move > world.time)
+			return
+		if( KB )
+			if(draw_from_slot_if_possible( KB ))
+				next_move = world.time + 3
+				return
+		for(var/slot in SLOT_DRAW_ORDER)
+			if(draw_from_slot_if_possible(slot))
+				next_move = world.time + 3
+				return
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:77]: magic numbers

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 8]
```diff
@@ -31,6 +31,63 @@
...
+	else
+		if(s_active && s_active.can_be_inserted(I))
+			s_active.handle_item_insertion(I, FALSE, src)
+			return
+		if( KB )
+			if(equip_to_slot_if_possible(I, KB , FALSE, FALSE, FALSE))
+				return
+		if(!equip_to_appropriate_slot(I, FALSE))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:85]: Use && as needed

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by TiviPlus** [comment]
> do not merge. in understand. but why the do not TM?

DNM is for splinter before I can comment
DNTM is because keybinds may nt be testmerged

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 2 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:60]: literally copy from do_quick_equip so.. idk . but sure.

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 7 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:77]: ?

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 4 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:70]: idk just changed this var a lot so i think i left a space there. well not anymore

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 3 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm]: ? huh

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 5 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm]: huh.. sure i guess

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 9]
```diff
@@ -31,6 +31,77 @@
...
+		if(!KB)
+			return
+
+		if(equip_to_slot_if_possible(I,KB, FALSE, FALSE, FALSE) && !equip_to_appropriate_slot(I, FALSE))
+			return
+
+		if(hand)
+			update_inv_l_hand(FALSE)
+		else
+			update_inv_r_hand(FALSE)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:104]:
```suggestion
	if(s_active?.can_be_inserted(I))
		s_active.handle_item_insertion(I, FALSE, src)
		return

	if(!KB)
		return

	if(equip_to_slot_if_possible(I,KB, FALSE, FALSE, FALSE) && !equip_to_appropriate_slot(I, FALSE))
		return

	if(hand)
		update_inv_l_hand(FALSE)
	else
		update_inv_r_hand(FALSE)
```

**Review by SplinterGP — APPROVED** [review]
seems fine for me mostly. just alot of keybinds and stuff but no problem.

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 10]
```diff
@@ -1,5 +1,4 @@
 /mob/living/carbon/human/proc/do_quick_equip()
-	SIGNAL_HANDLER_DOES_SLEEP
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:2]: Do not remove signal handlers

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 11]
```diff
@@ -31,6 +30,78 @@
...
+	do_kb_equip(SLOT_IN_HOLSTER,COMSIG_KB_SLOT_BELT_HOLSTER)
+
+/mob/living/carbon/human/proc/do_kb_equip_slot_suitsholster()
+	do_kb_equip(SLOT_IN_S_HOLSTER,COMSIG_KB_SLOT_SLOT_S_HOLSTER)
+
+/mob/living/carbon/human/proc/do_kb_equip_slot_backholster()
+	do_kb_equip(SLOT_IN_B_HOLSTER,COMSIG_KB_SLOT_B_HOLSTER)
+
+/mob/living/carbon/human/proc/do_kb_equip(KB,COMSIG)
+	. = COMSIG //The return value must be a flag compatible with the signals triggering this.
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:70]: COMSIG what

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 12]
```diff
@@ -31,6 +30,78 @@
...
+			return FALSE
+		if(draw_from_slot_if_possible(KB))
+			next_move = world.time
+			return
+		for(var/slot in SLOT_DRAW_ORDER)
+			if(draw_from_slot_if_possible(slot))
+				next_move = world.time
+			return
+	else
+		if(s_active?.can_be_inserted(I))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:90]: early returns

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 13]
```diff
@@ -31,6 +30,78 @@
...
+			s_active.handle_item_insertion(I, FALSE, src)
+			return
+
+		if(!KB)
+			return FALSE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:95]: why are you returning a bool here but not anywhere else

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 14]
```diff
@@ -31,6 +30,78 @@
...
+	if(incapacitated() || lying_angle || istype(loc, /obj/vehicle/multitile/root/cm_armored))
+		return
+
+	var/obj/item/I = get_active_held_item()
+	if(!I)
+		if(!next_move > world.time)
+			return FALSE
+
+		if(!KB)
+			return FALSE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:81]: ditto

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 15]
```diff
@@ -11,11 +10,11 @@
 			return
 		if(client?.prefs?.preferred_slot)
 			if(draw_from_slot_if_possible(client.prefs.preferred_slot))
-				next_move = world.time + 3
+				next_move = world.time
 				return
 		for(var/slot in SLOT_DRAW_ORDER)
 			if(draw_from_slot_if_possible(slot))
-				next_move = world.time + 3
+				next_move = world.time
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:17]: what is this

File: code/modules/mob/living/carbon/human/inventory.dm [hunk 16]
```diff
@@ -31,6 +30,78 @@
...
+			return FALSE
+
+		if(!KB)
+			return FALSE
+		if(draw_from_slot_if_possible(KB))
+			next_move = world.time
+			return
+		for(var/slot in SLOT_DRAW_ORDER)
+			if(draw_from_slot_if_possible(slot))
+				next_move = world.time
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/human/inventory.dm:87]: ditto

File: code/__DEFINES/equipment.dm [hunk 17]
```diff
@@ -350,3 +350,4 @@
 #define ITEM_NOT_EQUIPPED 0
 #define ITEM_EQUIPPED_CARRIED 1 //To hands, a storage or the likes.
 #define ITEM_EQUIPPED_WORN 2 //Actually worn on the body.
+#define DELAY_KB_EQUIP 3
```
Comment (TiviPlus) [inline code/__DEFINES/equipment.dm:353]: this isnt used anywhere

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 13 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:95]: so it returns anything but the " . "

thats equal to the COMSIG of the original keybind

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 11 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:70]: its for the proc right above it. it changes based on what key calls it.

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 10 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:2]: but you said that handler. was "illegal" sooner

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 14 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:81]: ditto

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 15 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:17]: well removing the delay that was in drawing weapons #5807

File: code/__DEFINES/equipment.dm (hunk 17 above)
Comment (fernandos13) [reply code/__DEFINES/equipment.dm:353]: true lemme delete it. still trying to figure out the early return thing tho

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 13 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm:95]: also this was supposed to be the early return but it guess.. this is.. not?

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 12 above)
Comment (fernandos13) [reply code/modules/mob/living/carbon/human/inventory.dm]: early returned

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 11 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/inventory.dm:70]: Make args lowercase

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 14 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/inventory.dm:81]: pending

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 15 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/inventory.dm:17]: nonatomized keep this

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 10 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/inventory.dm:2]: I said adding new sleeping ones is not allowed 
do not remove signal handlers

File: code/modules/mob/living/carbon/human/inventory.dm (hunk 11 above)
Comment (spookydonut) [reply code/modules/mob/living/carbon/human/inventory.dm:70]: you arent even returning any flags

**Pull request closed without merging**